jtransc-rt/src/java/util/Timer.java

```diff
@@ -1,588 +1,298 @@
-            cancelled = true;
-            tasks.reset();
-            this.notify();
-        }
+             if (result != 0)
+                 queue.heapify();
+         }
 
-        public int purge() {
-            if (tasks.isEmpty()) {
-                return 0;
-            }
-            // callers are synchronized
-            tasks.deletedCancelledNumber = 0;
-            tasks.deleteIfCancelled();
-            return tasks.deletedCancelledNumber;
-        }
+         return result;
+     }
+}
 
-    }
+class TimerThread extends Thread {
+	boolean newTasksMayBeScheduled = true;
+	private TaskQueue queue;
+
+	TimerThread(TaskQueue queue) {
+		this.queue = queue;
+	}
+
+	public void run() {
+		try {
+			mainLoop();
+		} finally {
+			// Someone killed this Thread, behave as if Timer cancelled
+			synchronized (queue) {
+				newTasksMayBeScheduled = false;
+				queue.clear();  // Eliminate obsolete references
+			}
+		}
+	}
+
+	private void mainLoop() {
+		while (true) {
+			try {
+				if (!newTasksMayBeScheduled) {
+					return;
+				}
+
+				boolean taskFired = false;
+				TimerTask task = null;
+
+				synchronized (queue) {
+					if (!queue.isEmpty()) {
+						long currentTime, executionTime;
+						task = queue.getMin();
+						synchronized (task.lock) {
+							if (task.state == TimerTask.CANCELLED) {
+								queue.removeMin();
+								continue;  // No action required, poll queue again
+							}
+							currentTime = System.currentTimeMillis();
+							executionTime = task.nextExecutionTime;
+
+							if (taskFired = (executionTime <= currentTime)) {
+								if (task.period == 0) { // Non-repeating, remove
+									queue.removeMin();
+									task.state = TimerTask.EXECUTED;
+								} else { // Repeating task, reschedule
+									queue.rescheduleMin(task.period < 0 ? currentTime - task.period : executionTime + task.period);
+								}
+							}
+						}
+					}
+				}
+
+				if (!taskFired) { // Task hasn't yet fired; wait
+					Thread.sleep(50);
+				} else if (task != null) {
+					task.run();
+				}
+			} catch (InterruptedException e) {
+			}
+		}
+	}
+}
 
-    private static final class FinalizerHelper {
-        private final TimerImpl impl;
+class TaskQueue {
 
-        FinalizerHelper(TimerImpl impl) {
-            this.impl = impl;
-        }
+    private TimerTask[] queue = new TimerTask[128];
+    private int size = 0;
 
-        @Override protected void finalize() throws Throwable {
-            try {
-                synchronized (impl) {
-                    impl.finished = true;
-                    impl.notify();
-                }
-            } finally {
-                super.finalize();
-            }
-        }
+    int size() {
+        return size;
     }
 
-    private static long timerId;
+    void add(TimerTask task) {
+        // Grow backing store if necessary
+        if (size + 1 == queue.length)
+            queue = Arrays.copyOf(queue, 2*queue.length);
```

Comment:
Since those are not integers or other primitives, why not to use an ArrayList instead? That way you would avoid having to do this manually